pkg/cloudprovider/providers/gce/gce_disks.go

```diff
@@ -278,49 +374,78 @@ func (gce *GCECloud) GetAutoLabelsForPD(name string, zone string) (map[string]st
-		// However, wherever possible the zone should be passed (and it is passed
-		// for most cases that we can control, e.g. dynamic volume provisioning)
+		// For regional PDs this is fine, but for zonal PDs we would like as far
+		// as possible to avoid this case, because GCE doesn't guarantee that
+		// volumes are uniquely named per region, just per zone. However,
+		// creation of GCE PDs was originally done only by name, so we have to
+		// continue to support that.
+		// However, wherever possible the zone should be passed (and it is
+		// passed for most cases that we can control, e.g. dynamic volume
+		// provisioning).
 		disk, err = gce.GetDiskByNameUnknownZone(name)
 		if err != nil {
 			return nil, err
 		}
-		zone = disk.Zone
 	} else {
 		// We could assume the disks exists; we have all the information we need
 		// However it is more consistent to ensure the disk exists,
 		// and in future we may gather addition information (e.g. disk type, IOPS etc)
-		disk, err = gce.getDiskByName(name, zone)
+		zoneSet, err := volumeutil.ZonesToSet(zone)
 		if err != nil {
-			return nil, err
+			glog.Warningf("Failed to parse zone field: %q. Will use raw field.", zone)
 		}
-	}
 
-	region, err := GetGCERegion(zone)
-	if err != nil {
-		return nil, err
+		if len(zoneSet) > 1 {
+			// Regional PD
+			disk, err = gce.getRegionalDiskByName(name)
+			if err != nil {
+				return nil, err
+			}
+		} else {
+			// Zonal PD
+			disk, err = gce.getDiskByName(name, zone)
+			if err != nil {
+				return nil, err
+			}
+		}
 	}
 
-	if zone == "" || region == "" {
+	labels := make(map[string]string)
+	switch zoneInfo := disk.ZoneInfo.(type) {
+	case singleZone:
+		if zoneInfo.zone == "" || disk.Region == "" {
+			// Unexpected, but sanity-check
+			return nil, fmt.Errorf("PD did not have zone/region information: %v", disk)
+		}
+		labels[kubeletapis.LabelZoneFailureDomain] = zoneInfo.zone
+		labels[kubeletapis.LabelZoneRegion] = disk.Region
+	case multiZone:
+		if zoneInfo.replicaZones == nil || zoneInfo.replicaZones.Len() <= 0 {
+			// Unexpected, but sanity-check
+			return nil, fmt.Errorf("PD is regional but does not have any replicaZones specified: %v", disk)
+		}
+		labels[kubeletapis.LabelZoneFailureDomain] = stringSetToCommaSeparatedString(zoneInfo.replicaZones)
+		labels[kubeletapis.LabelZoneRegion] = disk.Region
+	case nil:
+		// Unexpected, but sanity-check
+		return nil, fmt.Errorf("PD did not have ZoneInfo: %v", disk)
+	default:
 		// Unexpected, but sanity-check
-		return nil, fmt.Errorf("PD did not have zone/region information: %q", disk.Name)
+		return nil, fmt.Errorf("disk.ZoneInfo has unexpected type %T", zoneInfo)
 	}
 
-	labels := make(map[string]string)
-	labels[kubeletapis.LabelZoneFailureDomain] = zone
-	labels[kubeletapis.LabelZoneRegion] = region
-
 	return labels, nil
 }
 
+func stringSetToCommaSeparatedString(strSet sets.String) string {
+	return strings.Join(strSet.UnsortedList(), ",")
```

Comment:
Any reason to prefer unsorted?  The order will then be random

Reply:
Since we don't care about the resulting order, thought I'd optimize and avoid the extra sort operation.